Extended toric functionality

HereAround opened the pull request:
This is a first attempt towards enabling previously disabled functionality for toric varieties. Among others, this includes:
- FanOfVariety (and its convenience method Fan),
- Cox ring,
- Stanley-Reisner and irrelevant ideal,
- Class group,
- torus invariant divisor class group,
- character lattice,
- maps among the latter 3,
- more and refined constructors for toric divisors (DivisorOfCharacter, DivisorOfClass),
- support for isprime_divisor,
- extension and reorganization of tests,
- update of the documentation.

Please note that I have added
> using Oscar
using Test

to the test-file of ToricVarieties. This has the advantage that this test file can be executed individually, which allows for a (kind-of) speedy test on newly implemented functions regarding ToricVarieties.

Crucially, I also had to add the package `Combinatorics` as a dependency of `Oscar`. This is required for the computation of the Stanley-Reisner ideal (select k out of n - display all possibilities and iterate over them). For precisely, this is required for the computation of the `primitive_collections`, which is actually an attribute of a polyhedral fan. As of this writing, the implementation of this function is located within the ToricVarieties folder. Still, I assume (please correct/comment) that this must be moved to a suitable place in `Polyhedral Geometry`.

HereAround committed:
Add method affine_open_cover for normal toric varieties

HereAround committed:
Add method fan_of_variety for normal toric varieties

HereAround committed:
Use pm_ntv(v) for normal toric varieties

HereAround committed:
Update documentation of ToricVarieties

HereAround committed:
Add method torusinvariant_divisor_group for normal toric varieties

HereAround committed:
Add method character_lattice for normal toric varieties

HereAround committed:
Add method map_from_character_to_principal_divisors for normal toric varieties

HereAround committed:
Add map_from_principal_divisors_to_class_group for normal toric varieties

HereAround committed:
Add class_group for normal toric varieties

HereAround committed:
Add cox_ring for normal_toric_varieties

HereAround committed:
Add method list_of_variables_of_cox_ring for normal toric varieties

HereAround committed:
Change name: map_from_principal_divisors_to_class_group -> map_from_weil_divisors_to_class_group

HereAround committed:
Add implementation of Stanley-Reisner ideal

HereAround committed:
Add convenience method fan for fan_of_variety of normal toric varieties

HereAround committed:
Add function DivisorOfCharacter for normal toric varieties

HereAround committed:
Normalize whitespaces
Change order of arguments for ToricDivisors
Improve organization of tests

HereAround committed:
Add method isprime_divisor for ToricDivisors

thofma commented:
The dependency on Combinatorics.jl should be removed. The recent drama about some julia packages just shows that relying on external packages is not the best idea. The call `Combinatorics.combinations(v, d)` can just be replaced by `Oscar.Hecke.subsets(v, d)` (should just be `subsets(v, d)`, but I guess some import/export things are missing).

Some other random remark. Maybe we can get rid of the superfluous white spaces near parentheses? I think most (and hopefully soon all) parts of Oscar use `f(x)` instead of `f( x )`.

HereAround commented:
@thofma Thank you. Those are good points. I will update this PR accordingly.

HereAround committed:
Implement primitive_collections of polyhedral fans

lkastner commented on src/ToricVarieties/AttributesAndMethods.jl:
```diff
@@ -81,10 +81,383 @@ julia> dim(mori)
+function fan_of_variety( v::NormalToricVariety )
+    rays = Matrix{Int}(Oscar.Polymake.common.primitive(pm_ntv( v ).RAYS))
+    cones = v.polymakeNTV.MAXIMAL_CONES
+    return PolyhedralFan( rays, cones )
```
```suggestion
   return PolyhedralFan(pm_ntv(v));
```
At some point I had already changed it to the above. Maybe it got removed during migration? Nevertheless, we might already have computed a lot of information about the fan which would be discarded by reconstructing it. Also one would have to deal with lineality, and with the suggestion we don't need to think about those details.

HereAround replied:
I agree. In fact, I would like to commit your suggestion. Still, for some reason it seems to be outdated. I have not updated/pushed since early Monday morning in Europe. Not sure what happened. Can you somehow "refresh"?

lkastner commented on src/ToricVarieties/AttributesAndMethods.jl:
```diff
@@ -37,7 +37,7 @@ export dim_of_torusfactor
 """
 function euler_characteristic( v::AbstractNormalToricVariety )
-    f_vector = Vector{Int}(v.polymakeNTV.F_VECTOR)
+    f_vector = Vector{Int}(pm_ntv( v ).F_VECTOR)
```
I just thought about lineality. Do we need to take precautions here? If I remember correctly, polymake ignores the lineality.

HereAround replied:
First, I am happy with the suggested change. Again, for some reason this is marked as outdated. If you want me to merge your suggestion, please try to refresh. Or I will just add it/rebase.

As for the theory question - I am using theorem 12.3.9 in CLS. Thereby, for an n-dim toric variety, we merely have to count the number of n-dim cones in the fan. I was under the assumption that the f_vector created above has the property that f_vector[dim(v)] is exactly that number. Am I mistaken?

HereAround replied:
This concern may as well apply to the Betti numbers. As neither the euler_characteristic nor the Betti_numbers were changed in this PR (other than moving the code up and down in an effort to group/structure the code by content), maybe it would be best to address this in a separate issue/PR? What do you think?

HereAround committed:
Add method torusinvariant_prime_divisors for normal toric varieties

HereAround committed:
Organize implementation of operations for normal toric varieties by subjects

HereAround committed:
Organize documentation of method for normal toric varieties by subject

HereAround committed:
Implement star subdivison of ith maximal cone for polyhedral fans

HereAround committed:
Add method blowup_on_ith_minimal_torus_orbit for normal toric varieties

HereAround committed:
Implement Cartesian/direct product of two polyhedral fans

HereAround committed:
Implement direct product for normal toric varieties

HereAround committed:
Restructure documentation for toric divisors

HereAround committed:
Implement method coefficients tor toric divisors

HereAround committed:
Implement method to decide if a normal toric variety is the projective space
Minor bugfix

HereAround commented:
@thofma 
- I was indeed able to remove the dependency on `Combinatorics`, as you suggested.
- Whitespaces were addressed in all files which concern `ToricVarieties.` I hope I did not miss instances.

Further changes/additions:
- I moved the computation of `primitive_collections` to the polyhedral geometry.
- I also added direct products and star subdivisions of polyhedral fans.
- Correspondingly, support for blowups and direct products of toric varieties is provided.
- A few minor functional extensions (e.g. access to the coefficients of a toric divisor).
- Restructured, extended and improved the test file and documentation.

One issue is, that the Cox ring (i.e. coordinate ring) of a toric variety is in general graded by a f.p. Z-module. Still, I was only able to create Z-graded polynomial rings. Am I missing the corresponding constructors? I would also like to mention that I only managed to created these rings over the rationals, not over the complex numbers.

thofma commented:
Thanks @HereAround. Regarding the grading using Z-modules, I think you can do this if you create the module as an abelian group. There should be some examples in the corresponding test files, mpoly-graded or something like that. I am not at my computer, so can't easily check right now.

fieker commented:
You can create arbitrarily graded rings:
```
A = abelian_group([2, 3, 0])
Qx, x = PolynomialRing(QQ, :x=>1:3)
R = grade(Qx, [A[1], A[2], A[3]])
```

lkastner commented on src/ToricVarieties/NormalToricVarieties/constructors.jl:
```diff
@@ -119,22 +119,22 @@ end
         RAYS = Oscar.matrix_for_polymake(rays),
         MAXIMAL_CONES = Incidence,
-    )
+   )
```
What do you think about getting rid of this constructor? My point is: We have `NormalToricVariety(PolyhedralFan)` and the `PolyhedralFan` constructor is similar, although it is more flexible in terms of the input. If we want to keep this constructor, we should make it more flexible, maybe along the lines of `src/Polytopes/PolyhedralFan/constructors.jl`:
```
function PolyhedralFan(Rays::Union{Oscar.MatElem,AbstractMatrix}, Incidence::IncidenceMatrix)
function PolyhedralFan(Rays::Union{Oscar.MatElem,AbstractMatrix}, LS::Union{Oscar.MatElem,AbstractMatrix}, Incidence::IncidenceMatrix)
```

HereAround replied:
I agree.

HereAround replied:
I have removed this constructor in c40bafdd610de3ce815f2c9fdd20c56c8abd45a8.

lkastner commented on src/Polytopes/PolyhedralFan/properties.jl:
```diff
@@ -361,3 +361,185 @@ true
+    
+    # return the new fan
+    return PolyhedralFan(newrays, IncidenceMatrix(maxcones))
+    
```
```suggestion
    return PolyhedralFan(Polymake.fan.product(pm_fan(PF1), pm_fan(PF2)))
```
@benlorenz just showed me, that "cartesian" is not in the name of the function in polymake. But maybe this just works? We think it is the same.

HereAround replied:
If this is already in polymake, then by all means let us use it.

HereAround commented:
> You can create arbitrarily graded rings:
> 
> ```
> A = abelian_group([2, 3, 0])
> Qx, x = PolynomialRing(QQ, :x=>1:3)
> R = grade(Qx, [A[1], A[2], A[3]])
> ```

Thank you @fieker! This is exactly what I have been looking for. I will update this PR accordingly.

HereAround committed:
Use polymake function to compute product of polyhedral fans

Co-authored-by: Lars Kastner <lkastner@users.noreply.github.com>

HereAround commented on src/Polytopes/PolyhedralFan/properties.jl:
```diff
@@ -361,3 +361,185 @@ true
+"""
+function primitive_collections(PF::PolyhedralFan)
+    # collect data
+    cones = [findall(x->x!=0, l) for l in eachrow(pm_fan(PF).MAXIMAL_CONES)]
```
Ok. Maybe I was confused by this.

Still - you are right. We should not throw away the first ray, be it labeled as 0th or 1st. And indeed, I had no intention of doing this.

I have a vague recollection that the output of polymake.jl was recently changed to address exactly this offset. Am I mistaken or are the rays now (at least in the output of polymake.jl) labeled from 1 onwards, rather than 0? Can you please confirm/falsify? Maybe this is the reason why this did not trigger any error during the tests that I conducted?

fingolfin commented on src/Polytopes/PolyhedralFan/properties.jl:
```diff
@@ -361,3 +361,149 @@ true
+    all_points = [i for i in 1 : pm_fan(PF).N_RAYS]
+    d_max = maximum([length(i) for i in cones]) + 1
+    # identify and return the primitive collections
+    collections = Vector{Int}[];
```
```suggestion
    collections = Vector{Int}[]
```

HereAround replied:
Thank you @fingolfin .

fingolfin commented on src/Polytopes/PolyhedralFan/properties.jl:
```diff
@@ -361,3 +361,149 @@ true
+    d_max = maximum([length(i) for i in cones]) + 1
+    # identify and return the primitive collections
+    collections = Vector{Int}[];
+    for d in 1 : d_max
```
```suggestion
    for d in 1:d_max
```

HereAround replied:
Thank you @fingolfin .

fingolfin commented on src/Polytopes/PolyhedralFan/properties.jl:
```diff
@@ -361,3 +361,149 @@ true
+    # identify and return the primitive collections
+    collections = Vector{Int}[];
+    for d in 1 : d_max
+        checked  = Vector{Int}[];
```
```suggestion
        checked  = Vector{Int}[]
```

HereAround replied:
Thank you @fingolfin .

fingolfin commented on src/Polytopes/PolyhedralFan/properties.jl:
```diff
@@ -361,3 +361,149 @@ true
+    for d in 1 : d_max
+        checked  = Vector{Int}[];
+        for cone in cones
+            if d <= length(cone)
```
This code nests 8-9 levels, that's deep, perhaps too deep. Makes it harder to read, but also to follow. Personally, I'd do the following (and then of course get rid of the matching `end` and change the indentation)
```suggestion
            d <= length(cone) || continue
```

HereAround replied:
Thank you for the suggestion @fingolfin . I'll take care of `end` and the indentation.

fingolfin replied:
OK. Seems this has not yet happened, which causes the CI failures

fingolfin commented on src/Polytopes/PolyhedralFan/properties.jl:
```diff
@@ -361,3 +361,149 @@ true
+                    scanner = setdiff(all_points, I_minus_j)
+                    for j in scanner
+                        I = vcat(I_minus_j, [j])
+                        if (I in checked) == false
```
Similar to above:
```suggestion
                       I in checked && continue
```

HereAround replied:
Thank you @fingolfin .

fingolfin commented on src/Polytopes/PolyhedralFan/properties.jl:
```diff
@@ -361,3 +361,149 @@ true
+                        if (I in checked) == false
+                            push!(checked, I)
+                            # (1) I is contained in the primitive collections iff it is not contained in any cone
+                            if !(0 in [!issubset(Set(I), Set(test_cone)) for test_cone in cones])
```
Does one really need the `Set` calls there? If so, maybe `Set(I)` should be pulled out there and stored in a separate var, so it isn't recomputed on each iteration over `cones`...

I think you are using here that `0 == false`, right? So we can simplify this to
```suggestion
                            if !(true in [issubset(Set(I), Set(test_cone)) for test_cone in cones])
```
And using `any` I guess one gets something like this (double check, I am super tired)
```suggestion
                            if !any(test_cone -> issubset(Set(I), Set(test_cone)), cones)
```
Perhaps it'd be nicer with `all`, though, a matter of tates?

HereAround replied:
Yes, this is based on `0 == false`. With `any` works as well and it also appears to be working without `Set`.

Thank you for the suggestion @fingolfin .

fingolfin commented on src/Polytopes/PolyhedralFan/properties.jl:
```diff
@@ -361,3 +361,149 @@ true
+                            # (1) I is contained in the primitive collections iff it is not contained in any cone
+                            if !(0 in [!issubset(Set(I), Set(test_cone)) for test_cone in cones])
+                                # (2) I is generator of the primitive collections iff primitive_collections does not contain a "smaller" generator
+                                if !(0 in [!issubset(Set(prim), Set(I)) for prim in collections])
```
See above

HereAround replied:
These simplifications should be fully addressed in 9a51b2bf2c68f202be632d51e0915b530e902bfa , 3037b67b1d1651f7e49172d9d73950a9caf4651b and 62f1b6307c13ab8c0cb60d2a82b9d0a673da0beb.

fingolfin commented on src/Polytopes/PolyhedralFan/properties.jl:
```diff
@@ -361,3 +361,149 @@ true
+    
+    # check if n-th maximal cone does exist
+    if length(maxcones) < n
+        throw(ArgumentError("Cannot subdivide maximal cone " * string(n) * " as it does not exist!"))
```
```suggestion
        throw(ArgumentError("Cannot subdivide maximal cone $n as it does not exist!"))
```

HereAround replied:
Thank you @fingolfin .

fingolfin commented on src/Polytopes/PolyhedralFan/properties.jl:
```diff
@@ -361,3 +361,149 @@ true
+        for j in 1:length(rays[1])
+            gens[i,j] = rays[maxcones[n][i]][j]
+        end
+    end
```
```suggestion
    for i in 1:length(maxcones[n]), j in 1:length(rays[1])
        gens[i,j] = rays[maxcones[n][i]][j]
    end
```

HereAround replied:
Thank you @fingolfin .

fingolfin commented on src/Polytopes/PolyhedralFan/properties.jl:
```diff
@@ -361,3 +361,149 @@ true
+        end
+    end
+    cone = Oscar.positive_hull(gens)
+    if issmooth(PolyhedralFan(cone)) == false
```
A matter of taste, but I find this easier to process mentally:
```suggestion
    if !issmooth(PolyhedralFan(cone))
```
But again, subjective, feel free to ignore

HereAround replied:
I appreciate that change. Thank you @fingolfin .

fingolfin commented on src/Polytopes/PolyhedralFan/properties.jl:
```diff
@@ -361,3 +361,149 @@ true
+    push!(indices, newindex)
+    d = dim(cone)
+    test_cones = Oscar.Hecke.subsets(indices,d)
+    newmaxcones = Vector{Int}[];
```
```suggestion
    newmaxcones = Vector{Int}[]
```

HereAround replied:
Thank you @fingolfin .

fingolfin reviewed (approved):
Fine in general, I just left some nitpicks

fingolfin commented on test/ToricVarieties/runtests.jl:
```diff
@@ -1,92 +1,225 @@
-    p = polyhedron( D )
-    @test dim( p ) == 0
-    @test ambient_dim( p ) == 2
+v = H5 * P2;
```
```suggestion
v = H5 * P2
```

HereAround replied:
Thank you @fingolfin .

fingolfin commented on src/Polytopes/PolyhedralFan/properties.jl:
```diff
@@ -361,3 +361,149 @@ true
+        end
+    end
+    # return the computed primitive collections
+    return collections;
```
```suggestion
    return collections
```

HereAround replied:
Thank you @fingolfin .

HereAround committed:
Remove unnecessary ";"

Co-authored-by: Max Horn <max@quendi.de>

HereAround committed:
Normalize whitespaces

Co-authored-by: Max Horn <max@quendi.de>

HereAround committed:
Remove unnecessary ";"

Co-authored-by: Max Horn <max@quendi.de>

HereAround committed:
Include variable value in error message more efficiently

Co-authored-by: Max Horn <max@quendi.de>

HereAround committed:
Simplify code for 2-nested loop

Co-authored-by: Max Horn <max@quendi.de>

HereAround committed:
Remove unnecessary ";"

Co-authored-by: Max Horn <max@quendi.de>

HereAround committed:
Remove unnecessary ";"

Co-authored-by: Max Horn <max@quendi.de>

HereAround committed:
Remove unnecessary ";"

Co-authored-by: Max Horn <max@quendi.de>

HereAround committed:
Improve readability of code by changing "( ) == false" to "!()"

Co-authored-by: Max Horn <max@quendi.de>

HereAround committed:
Remove one constructor for toric varieties, as suggested by @lkastner

HereAround committed:
Simplify code for primitive_collections

Co-authored-by: Max Horn <max@quendi.de>

HereAround committed:
Reduce nesting level for primitive_collection algorithm

Co-authored-by: Max Horn <max@quendi.de>

HereAround committed:
Reduce nesting level for primitive_collection algorithm

Co-authored-by: Max Horn <max@quendi.de>

fingolfin commented:
Tests are failing.

Assuming they are fixed, should this be merged, or do you plan to do more work on it, @HereAround ?

HereAround commented:
I am still working on this. Specifically, I wish to follow the suggestion of @fieker and use `abelian_group`. For reasons of consistency, this needs addressing in a number of places. To make this PR nice, I am currently performing an involved rebase to address this. Once complete, this should also address the failed tests.

HereAround committed:
For convenience, allow to only run ToricVarieties tests

HereAround commented:
@fieker I am using `abelian_group` in the following commits
- character lattice of toric varieties: 4a61e6824549140dd13da19cc0f3f699f8d7ce92
- torusinvaraiant_divisor group: e225d42af5be1c13a34d39cf27096fa9035d716e
- class_group: b3d8c24213768703479e461f5d6965280d85388f
- maps among these: 0e1f368c7edb430509edda7fd105f87adc1d7a2c, a45c91f663d909c6b0b9ae3252487e86d19840d2

Eventually, this leads  to the desired grading of the Cox ring in bfde193229c330db646b1e2e88ac03eec78c1e2f.

I do not plan any major additions (in fact, this PR is probably larger than it should be for an easy reviewing process). If everybody is happy, please consider merging. Thank you.

fingolfin commented on src/Polytopes/PolyhedralFan/properties.jl:
```diff
@@ -361,3 +361,145 @@ true
+        for j in 1:length(rays[i])
+            newrays[i,j] = rays[i][j]
+        end
+    end
```
```suggestion
    for i in 1:length(rays), j in 1:length(rays[i])
        newrays[i,j] = rays[i][j]
    end
```

HereAround replied:
Thank you @fingolfin .

fingolfin commented on src/Polytopes/PolyhedralFan/properties.jl:
```diff
@@ -361,3 +361,145 @@ true
+    
+    # compute new rays to be added from star subdivision
+    newindex = size(rays,1) + 1
+    newray = [sum(gens[:,i]) for i in 1 : size(gens,2)]
```
```suggestion
    newray = [sum(gens[:,i]) for i in 1:size(gens,2)]
```
Same in other places, I'll not mark all of them

HereAround replied:
Ok. I'll search and replace these.

HereAround replied:
Done in https://github.com/oscar-system/Oscar.jl/pull/755/commits/4c1009758bd1b9fd17a90250fbacc3e4ca7ab614.

fingolfin commented on src/ToricVarieties/AttributesAndMethods.jl:
````diff
@@ -1,50 +1,345 @@
+ x3
+```
+"""
+function list_of_variables_of_cox_ring(v::AbstractNormalToricVariety)
````
This name should perhaps be reconsidered... Or perhaps this should indeed be dropped? After all `gens(cox_ring(v))` is both shorter than `list_of_variables_of_cox_ring(v)` and easier to "guess" for a general Oscar user

HereAround replied:
I believe you are right. This can be dropped. Will take care of this momentarily.

HereAround replied:
Done in https://github.com/oscar-system/Oscar.jl/pull/755/commits/b47a618c3a72aa923842153cd5a4e1a2b92fa051.

fingolfin commented on src/ToricVarieties/AttributesAndMethods.jl:
```diff
@@ -1,50 +1,345 @@
+
+@doc Markdown.doc"""
+    stanley_reisner_ideal(v::NormalToricVariety)
+Computes the Stanley-Reisner ideal of a normal toric variety `v`.
```
```suggestion

Computes the Stanley-Reisner ideal of a normal toric variety `v`.
```

HereAround replied:
Thank you @fingolfin .

fingolfin commented on src/ToricVarieties/AttributesAndMethods.jl:
```diff
@@ -1,50 +1,345 @@
+julia> p2 = toric_projective_space(2)
+A normal toric variety corresponding to a polyhedral fan in ambient dimension 2
+
+julia> length(stanley_reisner_ideal(P2).gens)
```
In general it's a bad idea to access struct members directly, this exposes implementation details. Better to do this (assuming it works; perhaps a suitable `gens` method needs to be installed first)
```suggestion
julia> length(gens(stanley_reisner_ideal(P2)))
```
Though actually the Oscar-ish way to do that would be to use `ngens`:
```suggestion
julia> ngens(stanley_reisner_ideal(P2))
```

HereAround replied:
Thank you for the suggestion @fingolfin. `ngens(stanley_reisner_ideal(P2))` does indeed exist. So I'll go with this change.

fingolfin commented on src/ToricVarieties/AttributesAndMethods.jl:
```diff
@@ -1,50 +1,345 @@
+"""
+function stanley_reisner_ideal(v::AbstractNormalToricVariety)
+    collections = primitive_collections(fan_of_variety(v))
+    SR_generators = []
```
Not type stable...

HereAround replied:
What do you think about https://github.com/oscar-system/Oscar.jl/pull/755/commits/bb10df654d9ea51de083e5fb55d7069c9deef513?

fingolfin commented on src/ToricVarieties/AttributesAndMethods.jl:
```diff
@@ -1,50 +1,345 @@
+           buffer = buffer * vars[I[k]]
+        end
+        push!(SR_generators, buffer)
+    end
```
Could perhaps be rewritten as
```suggestion
    for I in collections
        push!(SR_generators, prod(vars[I]))
    end
```
or even
```suggestion
    SR_generators = [prod(vars[I]) for I in collections]
```

HereAround replied:
Yes, that is a very good simplification. Thank you @fingolfin .

fingolfin commented on src/ToricVarieties/AttributesAndMethods.jl:
```diff
@@ -1,50 +1,345 @@
+    maximal_cones = []
+    for c in max_cones
+        buffer = Vector{Int}(undef, n_ray)
+        buffer = fill(0, n_ray)
```
```suggestion
        buffer = zeroes(Int, n_ray)
```

HereAround replied:
Thank you @fingolfin .

fingolfin commented on src/ToricVarieties/AttributesAndMethods.jl:
```diff
@@ -1,50 +1,345 @@
+            monom = monom * indeterminates[j]^(1 - maximal_cones[i][j])
+        end
+        push!(gens, monom)
+    end
```
Both `gens` and `monom` are not type stable (though it may not matter in this code *shrug). Also, another superfluous `;`.

HereAround replied:
Regarding type stability - does https://github.com/oscar-system/Oscar.jl/pull/755/commits/bb10df654d9ea51de083e5fb55d7069c9deef513 seem to fix this?

HereAround replied:
I removed ";" (among others) in https://github.com/oscar-system/Oscar.jl/pull/755/commits/4c1009758bd1b9fd17a90250fbacc3e4ca7ab614.

fingolfin commented on src/ToricVarieties/AttributesAndMethods.jl:
````diff
@@ -1,50 +1,345 @@
+```
+"""
+function map_from_character_to_principal_divisors(v::AbstractNormalToricVariety)
+    matrix = Matrix{Int}(Oscar.Polymake.common.primitive(pm_ntv(v).RAYS))
````
We are inside `Oscar`, so you can drop `Oscar.` (and should do it elsewhere, too)

HereAround replied:
Sorry about that. Will search and replace these.

HereAround replied:
Done in https://github.com/oscar-system/Oscar.jl/pull/755/commits/4c1009758bd1b9fd17a90250fbacc3e4ca7ab614.

fingolfin commented on src/ToricVarieties/AttributesAndMethods.jl:
```diff
@@ -1,50 +1,345 @@
+"""
+function map_from_character_to_principal_divisors(v::AbstractNormalToricVariety)
+    matrix = Matrix{Int}(Oscar.Polymake.common.primitive(pm_ntv(v).RAYS))
+    abstract_matrix = AbstractAlgebra.matrix(ZZ, size(matrix,2), size(matrix,1), vec(matrix))
```
Perhaps rename the variable `matrix` to e.g. `mat` and then change `AbstractAlgebra.matrix` to `matrix`?

Also, I am somewhat surprised that we have to go from whatever Polymake returns to `Matrix{Int}` and then to an Oscar matrix. But perhaps you only do it to get a more efficient transpose?

HereAround replied:
I am happy to rename these variables. Allow me to postpone this until I know if/what changes are required to address your second point.

`pm_ntv(v).RAYS` is not accepted as third argument for the group homomorphism in line 258. This is not unexpected, but rather desired. Namely, there we construct a map of (free) `Z`-modules whereas `pm_ntv(v).RAYS` is of type `pm::Matrix<pm::Rational>`.

My next guess was to use
> matrix = Matrix{Int}(Oscar.Polymake.common.primitive(pm_ntv(v).RAYS))
as third argument. Still, this `pm::Matrix<pm::Integer>` is not accepted as third argument.

At that point I went looking into the documentation of `abelian_group`: https://oscar-system.github.io/Oscar.jl/dev/Hecke/abelian/introduction/#abelian_group-Tuple{Matrix{fmpz}}. I vaguely recall that a group homomorphism constructor was described somewhere, with abstract matrix as third matrix. Right now, I cannot seem to find this. Still, abstract matrices appear frequently in the constructors of these groups. Hence, trying an abstract_matrix was a natural next guess. And this happened to work.

fingolfin replied:
It's possible that a method is missing there; or maybe it is not and what you do is the optimal way to handle this. I don't know... perhaps @thofma or @fieker have more to say, though.

But from my POV it is OK to merge this function as-is (well, perhaps after some renaming -- `abstract_matrix` is a matrix from AbstractAlgebra, but it is *not* of type `AbstractMatrix`, which is confusing :-)

HereAround replied:
The renaming has been done in https://github.com/oscar-system/Oscar.jl/pull/755/commits/e422bf2686ef420d5746418c25af4621bb8841dc.

fingolfin commented on src/ToricVarieties/AttributesAndMethods.jl:
```diff
@@ -1,50 +1,345 @@
+"""
+function torusinvariant_prime_divisors(v::AbstractNormalToricVariety)
+    ti_divisors = torusinvariant_divisor_group(v)
+    prime_divisors = []
```
```suggestion
    prime_divisors = ToricDivisor[]
```

HereAround replied:
Thank you @fingolfin .

fingolfin commented on src/ToricVarieties/AttributesAndMethods.jl:
```diff
@@ -1,50 +1,345 @@
+    ti_divisors = torusinvariant_divisor_group(v)
+    prime_divisors = []
+    for i in 1:rank(ti_divisors)
+        coeffs = fill(0, rank(ti_divisors))
```
```suggestion
        coeffs = zeros(Int, rank(ti_divisors))
```

HereAround replied:
Thank you @fingolfin .

fingolfin commented on src/ToricVarieties/AttributesAndMethods.jl:
```diff
@@ -138,3 +500,98 @@ export ith_betti_number
+function Base.:*(v::AbstractNormalToricVariety, w::AbstractNormalToricVariety)
+    return NormalToricVariety(fan_of_variety(v)*fan_of_variety(w))
+end
+export *
```
```suggestion
```

HereAround replied:
Thank you @fingolfin .

fingolfin commented on src/ToricVarieties/AttributesAndMethods.jl:
```diff
@@ -138,3 +500,98 @@ export ith_betti_number
+    w = [[Int(x) for x in transpose(g.coeff)] for g in gens(class_group(v))]
+    for i in 1:length(w)
+        g = w[i]
+        g = g[findall(x -> x!=0, g)]
```
Something like this may also work:
```suggestion
    for g in gens(class_group(v))
        g = [Int(x) for x in transpose(g.coeff) if !iszero(x)]
```
Actually, do you really need the call `transpose`?

HereAround replied:
This is a good suggestion. Also, I indeed think the transpose can be dropped.

HereAround replied:
Done in https://github.com/oscar-system/Oscar.jl/pull/755/commits/c6c9ab874dee31b54b37400769f98bc660ec0a0e.

fingolfin commented on src/ToricVarieties/AttributesAndMethods.jl:
```diff
@@ -138,3 +500,98 @@ export ith_betti_number
+    if irrelevant_ideal(v) != ideal(gens(cox_ring(v)))
+        return false
+    end
+    return true
```
```suggestion
    return irrelevant_ideal(v) == ideal(gens(cox_ring(v)))
```

HereAround replied:
Thank you @fingolfin .

fingolfin commented on src/ToricVarieties/NormalToricVarieties/constructors.jl:
```diff
@@ -171,56 +150,56 @@ export toric_projective_space
+function hirzebruch_surface(r::Int)
+    Rays = [1 0; 0 1; -1 r; 0 -1]
+    Cones = IncidenceMatrix([[1,2],[2,3],[3,4],[4,1]])
+    return NormalToricVariety(PolyhedralFan(Rays, Cones))
```
```suggestion
    rays = [1 0; 0 1; -1 r; 0 -1]
    cones = IncidenceMatrix([[1,2],[2,3],[3,4],[4,1]])
    return NormalToricVariety(PolyhedralFan(rays, cones))
```

HereAround replied:
Ok. I'll search and replace Rays by rays and Cones by cones.

HereAround replied:
Done in https://github.com/oscar-system/Oscar.jl/pull/755/commits/4901d082d87799a42fad578e7e105b45243e7cb1.

fingolfin commented on src/ToricVarieties/NormalToricVarieties/constructors.jl:
```diff
@@ -171,56 +150,56 @@ export toric_projective_space
-        Rays = [ 1 0; 0 1; -1 0; -1 -1 ]
-        Cones = [ [1,2],[2,3],[3,4],[4,1] ]
-        return NormalToricVariety( Rays, Cones )
+        Rays = [1 0; 0 1; -1 0; -1 -1]
```
should change Rays -> rays, Cones -> cones in this function, too

HereAround replied:
Ok. I'll search and replace Rays by rays and Cones by cones.

HereAround replied:
Done in https://github.com/oscar-system/Oscar.jl/pull/755/commits/4901d082d87799a42fad578e7e105b45243e7cb1.

fingolfin commented on src/ToricVarieties/ToricDivisors.jl:
```diff
@@ -40,11 +40,38 @@ export ToricDivisor
+"""
+function DivisorOfCharacter(v::AbstractNormalToricVariety, character::Vector{Int})
+    if length(character) != rank(character_lattice(v))
+        throw(ArgumentError("Character must consist of " * string(rank(character_lattice(v))) * " integers!"))
```
```suggestion
        throw(ArgumentError("Character must consist of $(rank(character_lattice(v))) integers!"))
```

HereAround replied:
Thank you @fingolfin .

fingolfin commented on src/ToricVarieties/ToricDivisors.jl:
```diff
@@ -40,11 +40,38 @@ export ToricDivisor
+        throw(ArgumentError("Character must consist of " * string(rank(character_lattice(v))) * " integers!"))
+    end
+    f = map_from_character_to_principal_divisors(v)
+    char = sum([character[i] * gens(domain(f))[i] for i in 1:length(gens(domain(f)))])
```
```suggestion
    char = sum(character .* gens(domain(f)))
```

HereAround replied:
Thank you @fingolfin .

fingolfin commented on src/ToricVarieties/ToricDivisors.jl:
```diff
@@ -224,6 +251,41 @@ isq_cartier(td::ToricDivisor) = pm_tdivisor(td).Q_CARTIER::Bool
+function isprime_divisor(td::ToricDivisor)
+    pm_divisor = td.polymake_divisor
+    coeffs = [Int(c) for c in Oscar.Polymake.common.primitive(pm_divisor.COEFFICIENTS)]
+    if (sum(coeffs) != 1)
```
```suggestion
    if sum(coeffs) != 1
```

HereAround replied:
Thank you @fingolfin .

fingolfin commented on src/ToricVarieties/ToricDivisors.jl:
```diff
@@ -224,6 +251,41 @@ isq_cartier(td::ToricDivisor) = pm_tdivisor(td).Q_CARTIER::Bool
+    if (all(y -> (y == 1 || y == 0), coeffs) == false)
+        return false
+    end
+    return true
```
```suggestion
    return all(y -> (y == 1 || y == 0), coeffs)
```

HereAround replied:
Thank you @fingolfin .

fingolfin reviewed (commented):
Some more nitpicks... a few of them apply to other parts in this PR, too

HereAround committed:
Simplify double-loop code

Co-authored-by: Max Horn <max@quendi.de>

HereAround committed:
Add empty line for codestyling of documentation code.

Co-authored-by: Max Horn <max@quendi.de>

HereAround committed:
Simplify code for generators of the Stanley-Reisner ideal

Co-authored-by: Max Horn <max@quendi.de>

HereAround committed:
Better code to tell if a divisor is prime.

Co-authored-by: Max Horn <max@quendi.de>

HereAround committed:
Remove unnecessary brackets

Co-authored-by: Max Horn <max@quendi.de>

HereAround committed:
Simplify code for character computation

Co-authored-by: Max Horn <max@quendi.de>

HereAround committed:
Include variable value more efficiently in error message

Co-authored-by: Max Horn <max@quendi.de>

HereAround committed:
Use ngens to access number of generators of ideal

Co-authored-by: Max Horn <max@quendi.de>

HereAround committed:
Simplify construction of vector with zero entries

Co-authored-by: Max Horn <max@quendi.de>

HereAround committed:
Initialize type of variable prime_divisors

Co-authored-by: Max Horn <max@quendi.de>

HereAround committed:
Simplify check if a toric variety is the projective space

Co-authored-by: Max Horn <max@quendi.de>

HereAround committed:
Better construction of zero vector

Co-authored-by: Max Horn <max@quendi.de>

HereAround committed:
Remove unncessary export

Co-authored-by: Max Horn <max@quendi.de>

HereAround committed:
Codestyling
" : " -> ":"
";"  ->  ""
"Oscar." -> ""

HereAround committed:
Improve method isprojective_space

HereAround committed:
Rename variables in ToricVarieties constructors:
Rays -> rays
Cones -> cones

HereAround committed:
Remove method list_of_variables_of_cox_ring

HereAround committed:
Address typestability in computation of irrelevant and SR-ideal

HereAround committed:
affine_open_cover now works for AbstractNormalToricVariety

fingolfin commented on src/Polytopes/PolyhedralFan/properties.jl:
```diff
@@ -361,3 +361,143 @@ true
+function starsubdivision(PF::PolyhedralFan, n::Int)
+    # extract defining information on the fan
+    maxcones = [findall(x->x!=0, l) for l in eachrow(pm_fan(PF).MAXIMAL_CONES)]
+    rays = [[Int(i) for i in r] for r in eachrow(Oscar.Polymake.common.primitive(pm_fan(PF).RAYS))]
```
Fundamental question: is it really safe to use `Int` everywhere? Is Polymake guaranteed to only produce output that fits in there, and that no overflow/underflow can happen?

Indeed, what type *does* `Oscar.Polymake.common.primitive` return, and which type do the coefficients have? @lkastner ?

Perhaps `fmpz` instead of `Int` here and in similar places would be better/safer?

Finally, taking one step back: shouldn't this code be rewritten to use the function `Oscar.rays(PF)`?

HereAround replied:
I can certainly look into rewriting this with `Oscar.rays(PF)`. Most likely, this is something that can/must be addressed more generally in `ToricVarieties`. I believe we often access the rays in `Polymake`.

Regarding the type - my understanding is that `pm_fan(FP)` returns a Polymake matrix with rational numbers as entries. Upon use of `primitive` it becomes a `pm::Matrix<pm::Integers>`. But your question is I believe concerned with `Int` vs `Int64` vs `fmpz`.

Long story short - I will be happy to use fmpz. @lkastner Opinions?

lkastner replied:
Oh, sorry, I didn't see this. polymake will probably produce its own version `Polymake.Integer`. These are arbitrary precision. I think `fmpz` is safer then. You can convert with `convert(fmpz, i)`. But maybe at this point we should revisit after the merge? Maybe it can even be done with `Polymake.Integer`? It is getting fed back to polymake in the end anyway.

fingolfin commented on src/Polytopes/PolyhedralFan/properties.jl:
```diff
@@ -361,3 +361,143 @@ true
+    end
+    cone = Oscar.positive_hull(gens)
+    if !issmooth(PolyhedralFan(cone))
+        throw(ArgumentError("Cannot subdivide maximal cone " * string(n) * " as it is not smooth!"))
```
In general, it is usually preferable to use string interpolation over string concatenation plus explicit calls to `string`
```suggestion
        throw(ArgumentError("Cannot subdivide maximal cone $n as it is not smooth!"))
```

thofma commented on test/ToricVarieties/runtests.jl:
```diff
@@ -1,92 +1,197 @@
+    @test rank(domain(map)) == 2
+    @test rank(codomain(map)) == 2
+    @test length(gens(codomain(map_from_weil_divisors_to_class_group(antv)))) == 2
+    @test length(gens(class_group(antv))) == 2
```
I don't think that `length(gens)/ngens` is an invariant of the abelian group. You could either test the isomorphism or `elementary_divisors(class_group(antv)) == [?, ?]`.

HereAround replied:
@thoma

Thank you for pointing this out.

Of course you are right. There are many different presentations for one and the same abelian group, which differ in their number of generators. I was under the impression that the presentation is not changed/simplified internally, so that the number of generators would be the ones in the input. Maybe this is wrong? Is there internally a simplification for the `abelian_group` which looks for a smaller presentation?

Ideally, my check would be to ask if the class_group is isomorphic to a certain group. E.g. for projective space, we know that the class_group is isomorphic to `Z`. So an ideal check would be something like `isequal(smaller_presentation(class_group(toric_projective_space(2))),Z)`. Is something along these lines possible?

Or maybe simpler and still providing a meaningful test - can I somehow ask if an `abelian_group` is free?

thofma replied:
> @thoma
> 
> Thank you for pointing this out.
> 
> Of course you are right. There are many different presentations for one and the same abelian group, which differ in their number of generators. I was under the impression that the presentation is not changed/simplified internally, so that the number of generators would be the ones in the input. Maybe this is wrong? Is there internally a simplification for the `abelian_group` which looks for a smaller presentation?

Yes, `ngens(A)` won't change for an already computed abelian group `A`, but the algorithm for computing the `cokernel` could change, such that a better (hopefully not worse) presentation is returned. The `cokernel` map does not guarentee specific presentations, just the properties of a cokernel.

You can ask for `S, = snf(::GrpAb)`, then `S` will be `Z/d_1 x ... x Z/d_n` with `d_i` dividing `d_i+1` (Smith normal form). I think this gives the minimal number of generators and is also canonical.

> Ideally, my check would be to ask if the class_group is isomorphic to a certain group. E.g. for projective space, we know that the class_group is isomorphic to `Z`. So an ideal check would be something like `isequal(smaller_presentation(class_group(toric_projective_space(2))),Z)`. Is something along these lines possible?
> 
> Or maybe simpler and still providing a meaningful test - can I somehow ask if an `abelian_group` is free?

You can call `isisomorphic(C, abelian_group([0])`. Unfortunately there is no `isfree(::GrpAb)`, but we should add it (https://github.com/thofma/Hecke.jl/issues/437). If there were, one could do `isfree(A) && rank(A) == 2`.

HereAround replied:
Thank you @thoma. That completely answers this. I will update this PR accordingly.

HereAround replied:
Adjusted (via `elementary_divisors`) in https://github.com/oscar-system/Oscar.jl/pull/755/commits/7133ef36fc663d8b594de1716d9bf41d7f062cfe.

lkastner committed:
Polytopes: Streamline star subdivision

HereAround committed:
Improve tests for class_group as suggested by @thofma

HereAround committed:
Rename variables as suggested by @fingolfin
